sgqlc/endpoint/websocket.py

```diff
@@ -1,36 +1,122 @@
+from typing import Dict, Any, Union, SupportsBytes, Iterable
+from websocket import WebSocket
+
 from sgqlc.endpoint.base import BaseEndpoint
 import websocket
 import uuid
 import json
 
 
+class CancelableResultGenerator:
+    '''
+    Generator of websocket subscription notification events that supports
+    sending a message to the server to cancel the subscription.
+    '''
+    def __init__(self, ws: WebSocket, query_id: str, should_close: bool):
+        '''
+        :param ws: websocket to get results from
+        :type ws: WebSocket
+
+        :param query_id: query_id to track
+        :type query_id: str
+
+        :param should_close: whether to close the socket after
+          completing iteration on the results
+        :type should_close: bool
+        '''
+        self.ws = ws
+        self.query_id = query_id
+        self.should_close = should_close
+
+    def __next__(self) -> dict:
+        response = json.loads(self.ws.recv())
+        if response['id'] != self.query_id:
+            raise ValueError(
+                f'Unexpected id {response["id"]} '
+                f'when waiting for query results'
+            )
+        if response['type'] == 'data':
+            return response['payload']
+        elif response['type'] == 'complete':
+            if self.should_close:
+                self.ws.close()
+            raise StopIteration()
+        else:
+            raise ValueError(f'Unexpected message {response} '
+                             f'when waiting for query results')
+
+    def __iter__(self):
+        return self
+
+    def cancel(self):
+        self.ws.send(json.dumps({'type': 'stop', 'id': self.query_id}))
+
+
 class WebSocketEndpoint(BaseEndpoint):
     '''
-    A synchronous websocket endpoint for graphql queries or subscriptions
+    A websocket endpoint for graphql queries or subscriptions.
+    Will return a generator of response objects when called.
     '''
-    def __init__(self, url, **ws_options):
+    def __init__(self, url: str = None, ws: WebSocket = None, **ws_options):
         '''
         :param url: ws:// or wss:// url to connect to
         :type url: str
 
+        :param ws: existing websocket connection to use instead of
+          creating a new one.  if passing an existing connection,
+          the calling code is responsible for both closing it and
+          handling the ``connection_init``/``connection ack``
+          handshake.
+        :type ws: WebSocket
+
         :param ws_options: options to pass to websocket.create_connection
         :type ws_options: dict
         '''
         self.url = url
+        self.ws = ws
```

Comment:
here you should assert `url` and `ws` are not given at the same time, creating `ws` if it's not provided.